tl.tl

```diff
@@ -11333,28 +11641,301 @@ do
+         end
+         local pat = b.tuple[2]
+         local orig_t = b.tuple[3]
+         local trepl = self:to_structural(orig_t)
+
+         local has_fourth = b.tuple[4]
+
+         local args: TupleType
+
+         if pat is StringType and pat.literal then
+            local st = pat.literal
+            local items, e = parse_pattern_string(node, st, true)
+
+            if e then
+               if items then
+                  -- just a warning
+                  self.errs:add_warning("hint", pat, e)
+               else
+                  return self.errs:invalid_at(pat, e)
+               end
+            end
+
+            local i1 = items[1]
+
+            -- we expect replarg to be one of:
+            -- string
+            -- {(items[1]):string}
+            -- function(table.unpack(items)): ((string|integer|number)...)
+
+            local replarg_type: Type
+
+            local expected_pat_return = a_union(node, {
+               a_type(node, "string", {}),
+               a_type(node, "integer", {}),
+               a_type(node, "number", {}),
+            })
+            if self:is_a(trepl, expected_pat_return) then
+               -- fine
+               replarg_type = expected_pat_return
+            elseif trepl is MapType then
+               replarg_type = a_map(node, i1, expected_pat_return)
+            elseif trepl is RecordLikeType then
+               if not (i1 is StringType) then
+                  self.errs:invalid_at(trepl, "expected a table with integers as keys")
+               end
+               replarg_type = a_map(node, i1, expected_pat_return)
+            elseif trepl is ArrayLikeType then
+               if not (i1 is IntegerType) then
+                  self.errs:invalid_at(trepl, "expected a table with strings as keys")
+               end
+               replarg_type = an_array(node, expected_pat_return)
+            elseif trepl is FunctionType then
+               local validftype = a_function(node, {
+                  min_arity = self.feat_arity and #items or 0,
+                  args = a_tuple(node, items),
+                  rets = a_vararg(node, { expected_pat_return }),
+               })
+               replarg_type = validftype
+            end
+            -- other types should be taken care of by the type_check_function_call below
+
+            if replarg_type then
+               args = a_tuple(node, {
+                  a_type(node, "string", {}),
+                  a_type(node, "string", {}),
+                  replarg_type,
+                  has_fourth and a_type(node, "integer", {}) or nil,
+               })
+            end
+         end
+
+         return (self:type_check_function_call(node, a, b, argdelta, args, nil))
+      end,
+   }
+
+   function TypeChecker:type_check_funcall(node: Node, a: Type, b: TupleType, argdelta?: integer): InvalidOrTupleType
+      argdelta = argdelta or 0
+
+      local special_tyck = special_functions[a.special_function_handler]
```

Comment:
Looks like the handling of `special_tyck` could move inside `type_check_function_call`, then it only needs to be handled once.

Reply:
Ah, nevermind, I see now why not!